File: 72_llm_market_simulation/python/simulation/engine.py
"""
Simulation Engine for LLM Market Simulation

Orchestrates the market simulation with multiple agents.
"""

from typing import Dict, List, Optional, Any
import numpy as np
from dataclasses import dataclass, field

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from market import MarketEnvironment, Order, OrderType, Side
from agents import BaseAgent, AgentDecision, ActionType


@dataclass
class SimulationResult:
    """
    Results from a simulation run

    Attributes:
        price_history: List of prices over time
        fundamental_history: List of fundamental values over time
        agent_performance: Dict mapping agent_id to performance metrics
        market_metrics: Market quality metrics
        trade_log: Log of all trades
    """
    price_history: List[float]
    fundamental_history: List[float]
    agent_performance: Dict[str, Dict[str, float]]
    market_metrics: Dict[str, float]
    trade_log: List[Dict[str, Any]] = field(default_factory=list)


class SimulationEngine:
    """
    Simulation Engine

    Orchestrates the market simulation by coordinating agents,
    processing orders, and collecting metrics.

    Examples:
        >>> engine = SimulationEngine(initial_price=100.0)
        >>> engine.add_agent(ValueInvestorAgent("value_1", 100000))
        >>> results = engine.run(num_steps=100)
    """

    def __init__(
        self,
        initial_price: float = 100.0,
        fundamental_value: Optional[float] = None,
        volatility: float = 0.02,
        tick_size: float = 0.01,
        random_seed: Optional[int] = None,
        verbose: bool = False
    ):
        """
        Initialize simulation engine

        Args:
            initial_price: Starting price
            fundamental_value: True fundamental value
            volatility: Daily volatility for fundamental value
            tick_size: Minimum price increment
            random_seed: Random seed for reproducibility
            verbose: Whether to print progress
        """
        self.market = MarketEnvironment(
            initial_price=initial_price,
            fundamental_value=fundamental_value,
            tick_size=tick_size,
            volatility=volatility,
            random_seed=random_seed
        )
        self.agents: List[BaseAgent] = []
        self.verbose = verbose
        self.rng = np.random.default_rng(random_seed)

    def add_agent(self, agent: BaseAgent):
        """
        Add an agent to the simulation

        Args:
            agent: Agent to add
        """
        self.agents.append(agent)
        self.market.register_agent(agent.agent_id, agent.initial_cash)

    def run(self, num_steps: int = 100) -> SimulationResult:
        """
        Run the simulation

        Args:
            num_steps: Number of time steps to simulate

        Returns:
            SimulationResult with all metrics and history
        """
        trade_log = []

        for step in range(num_steps):
            if self.verbose and step % 50 == 0:
                print(f"Step {step}/{num_steps}, Price: ${self.market.current_price:.2f}")

            # Get current market state
            state = self.market.get_state()

            # Each agent makes a decision
            decisions = []
            for agent in self.agents:
                decision = agent.make_decision(
                    current_price=state.current_price,
                    fundamental_value=state.fundamental_value,
                    price_history=state.price_history,
                    market_state={
                        "best_bid": state.best_bid,
                        "best_ask": state.best_ask,
                        "spread": state.spread,
                        "volume_24h": state.volume_24h,
                        "time_step": state.time_step
                    }
                )
                decisions.append((agent, decision))

            # Shuffle order to avoid first-mover advantage
            self.rng.shuffle(decisions)

            # Process decisions
            for agent, decision in decisions:
                if decision.action == ActionType.HOLD:
                    continue

                # Create order
                side = Side.BUY if decision.action == ActionType.BUY else Side.SELL
                order_type = OrderType.LIMIT if decision.order_type == "limit" else OrderType.MARKET

                order = Order(
                    order_type=order_type,
                    side=side,
                    quantity=decision.quantity,
                    price=decision.limit_price if order_type == OrderType.LIMIT else None,
                    agent_id=agent.agent_id
                )

                # Submit order
                result = self.market.submit_order(agent.agent_id, order)

                # Update agent portfolio
                if result.filled_quantity > 0:
                    agent.update_portfolio(
                        decision.action,
                        result.filled_quantity,
                        result.average_price
                    )

                    trade_log.append({
                        "step": step,
                        "agent_id": agent.agent_id,
                        "action": decision.action.value,
                        "quantity": result.filled_quantity,
                        "price": result.average_price,
                        "reasoning": decision.reasoning[:100]
                    })

            # Advance market
            self.market.step()

        # Collect results
        agent_performance = {}
        for agent in self.agents:
            final_value = agent.get_portfolio_value(self.market.current_price)
            returns = agent.get_return(self.market.current_price)

            agent_performance[agent.agent_id] = {
                "final_value": final_value,
                "return": returns,
                "return_pct": returns * 100,
                "num_trades": len(agent.trade_history),
                "final_holdings": agent.holdings,
                "final_cash": agent.cash
            }

        return SimulationResult(
            price_history=list(self.market.price_history),
            fundamental_history=list(self.market.fundamental_history),
            agent_performance=agent_performance,
            market_metrics=self.market.calculate_metrics(),
            trade_log=trade_log
        )

    def reset(self):
        """Reset simulation to initial state"""
        self.market.reset()
        for agent in self.agents:
            agent.reset()


if __name__ == "__main__":
    from agents import ValueInvestorAgent, MomentumTraderAgent, MarketMakerAgent

    print("Running Market Simulation Demo")
    print("=" * 50)

    # Create simulation
    engine = SimulationEngine(
        initial_price=100.0,
        fundamental_value=100.0,
        volatility=0.02,
        random_seed=42,
        verbose=True
    )

    # Add agents
    engine.add_agent(ValueInvestorAgent("value_1", initial_cash=100000))
    engine.add_agent(MomentumTraderAgent("momentum_1", initial_cash=100000))
    engine.add_agent(MarketMakerAgent("mm_1", initial_cash=100000))

    # Run simulation
    results = engine.run(num_steps=100)

    # Print results
    print("\n" + "=" * 50)
    print("SIMULATION RESULTS")
    print("=" * 50)

    print(f"\nFinal Price: ${results.price_history[-1]:.2f}")
    print(f"Final Fundamental: ${results.fundamental_history[-1]:.2f}")

    print("\nAgent Performance:")
    for agent_id, perf in results.agent_performance.items():
        print(f"  {agent_id}:")
        print(f"    Return: {perf['return_pct']:.2f}%")
        print(f"    Final Value: ${perf['final_value']:.2f}")
        print(f"    Trades: {perf['num_trades']}")

    print("\nMarket Metrics:")
    for metric, value in results.market_metrics.items():
        print(f"  {metric}: {value:.4f}")

File: 72_llm_market_simulation/rust_market_sim/src/simulation/engine.rs
//! Simulation Engine
//!
//! Orchestrates multi-agent market simulation.

use crate::agents::{Agent, AgentDecision, Action};
use crate::agents::base::MarketState;
use crate::market::{OrderBook, Order, OrderType, Side};
use rand::Rng;
use rand_distr::{Distribution, Normal};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Result for a single agent
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentResult {
    /// Agent ID
    pub agent_id: String,
    /// Agent type
    pub agent_type: String,
    /// Final cash balance
    pub final_cash: f64,
    /// Final shares
    pub final_shares: i64,
    /// Final portfolio value
    pub final_value: f64,
    /// Number of trades
    pub num_trades: u64,
    /// Return percentage
    pub return_pct: f64,
}

/// Simulation result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimulationResult {
    /// Price history
    pub price_history: Vec<f64>,
    /// Fundamental value history
    pub fundamental_history: Vec<f64>,
    /// Results per agent
    pub agent_results: HashMap<String, AgentResult>,
    /// Total trades executed
    pub total_trades: u64,
    /// Final price
    pub final_price: f64,
    /// Number of steps completed
    pub steps_completed: u64,
}

/// Simulation Engine
pub struct SimulationEngine {
    /// Order book
    order_book: OrderBook,
    /// Trading agents
    agents: HashMap<String, Box<dyn Agent>>,
    /// Agent order for round-robin execution
    agent_order: Vec<String>,
    /// Fundamental value
    fundamental_value: f64,
    /// Price volatility per step
    volatility: f64,
    /// Price history
    price_history: Vec<f64>,
    /// Fundamental history
    fundamental_history: Vec<f64>,
    /// Trade counter
    trade_count: u64,
    /// Current step
    current_step: u64,
    /// Initial price
    initial_price: f64,
}

impl SimulationEngine {
    /// Create a new simulation engine
    pub fn new(initial_price: f64, fundamental_value: f64, volatility: f64) -> Self {
        Self {
            order_book: OrderBook::new(initial_price),
            agents: HashMap::new(),
            agent_order: Vec::new(),
            fundamental_value,
            volatility,
            price_history: vec![initial_price],
            fundamental_history: vec![fundamental_value],
            trade_count: 0,
            current_step: 0,
            initial_price,
        }
    }

    /// Add an agent to the simulation
    pub fn add_agent(&mut self, agent: Box<dyn Agent>) {
        let id = agent.id().to_string();
        self.agent_order.push(id.clone());
        self.agents.insert(id, agent);
    }

    /// Get current market state for agents
    fn get_market_state(&self) -> MarketState {
        MarketState {
            current_price: self.order_book.last_price,
            best_bid: self.order_book.best_bid(),
            best_ask: self.order_book.best_ask(),
            price_history: self.price_history.clone(),
            fundamental_value: self.fundamental_value,
            step: self.current_step,
        }
    }

    /// Process an agent's decision
    fn process_decision(&mut self, agent_id: &str, decision: &AgentDecision) {
        if matches!(decision.action, Action::Hold) {
            return;
        }

        let side = match decision.action {
            Action::Buy => Side::Buy,
            Action::Sell => Side::Sell,
            Action::Hold => return,
        };

        let price = decision.limit_price.unwrap_or(self.order_book.last_price);

        let order = Order::new(
            agent_id.to_string(),
            side,
            decision.order_type,
            price,
            decision.quantity,
        );

        let result = self.order_book.submit_order(order);

        // Update agent positions for filled orders
        if result.filled_quantity > 0 {
            self.trade_count += 1;

            let agent = self.agents.get_mut(agent_id).unwrap();
            let value = result.average_price * result.filled_quantity as f64;

            match side {
                Side::Buy => {
                    agent.update_position(-value, result.filled_quantity as i64);
                }
                Side::Sell => {
                    agent.update_position(value, -(result.filled_quantity as i64));
                }
            }
        }
    }

    /// Update fundamental value with random walk
    fn update_fundamental(&mut self) {
        let mut rng = rand::thread_rng();
        let normal = Normal::new(0.0, self.volatility * 0.5).unwrap();
        let change = normal.sample(&mut rng);
        self.fundamental_value *= 1.0 + change;
        self.fundamental_history.push(self.fundamental_value);
    }

    /// Add noise to price (represents external factors)
    fn add_price_noise(&mut self) {
        let mut rng = rand::thread_rng();
        let normal = Normal::new(0.0, self.volatility).unwrap();
        let noise = normal.sample(&mut rng);

        // Price reverts slightly toward fundamental
        let reversion = (self.fundamental_value - self.order_book.last_price) * 0.01;

        let new_price = self.order_book.last_price * (1.0 + noise) + reversion;
        self.order_book.last_price = new_price.max(1.0); // Prevent negative prices
    }

    /// Run one simulation step
    fn step(&mut self) {
        self.current_step += 1;

        // Update fundamental value
        self.update_fundamental();

        // Get market state
        let state = self.get_market_state();

        // Each agent makes a decision
        let agent_ids: Vec<String> = self.agent_order.clone();
        for agent_id in agent_ids {
            let agent = self.agents.get_mut(&agent_id).unwrap();
            let decision = agent.make_decision(&state);
            drop(agent); // Release borrow

            self.process_decision(&agent_id, &decision);
        }

        // Add price noise and record
        self.add_price_noise();
        self.price_history.push(self.order_book.last_price);

        // Clear old orders periodically
        if self.current_step % 10 == 0 {
            self.order_book.clear();
        }
    }

    /// Run simulation for specified number of steps
    pub fn run(&mut self, num_steps: u64) -> SimulationResult {
        for _ in 0..num_steps {
            self.step();
        }

        self.get_results()
    }

    /// Run simulation with progress callback
    pub fn run_with_progress<F>(&mut self, num_steps: u64, mut callback: F) -> SimulationResult
    where
        F: FnMut(u64, f64),
    {
        for i in 0..num_steps {
            self.step();
            if i % 10 == 0 {
                callback(i, self.order_book.last_price);
            }
        }

        self.get_results()
    }

    /// Get simulation results
    fn get_results(&self) -> SimulationResult {
        let mut agent_results = HashMap::new();

        for (id, agent) in &self.agents {
            let final_value = agent.portfolio_value(self.order_book.last_price);
            let initial_value = self.initial_price * 100.0 + 100000.0; // Rough estimate

            agent_results.insert(id.clone(), AgentResult {
                agent_id: id.clone(),
                agent_type: agent.agent_type().to_string(),
                final_cash: agent.cash(),
                final_shares: agent.shares(),
                final_value,
                num_trades: 0, // Would need to track per agent
                return_pct: (final_value / initial_value - 1.0) * 100.0,
            });
        }

        SimulationResult {
            price_history: self.price_history.clone(),
            fundamental_history: self.fundamental_history.clone(),
            agent_results,
            total_trades: self.trade_count,
            final_price: self.order_book.last_price,
            steps_completed: self.current_step,
        }
    }

    /// Reset simulation
    pub fn reset(&mut self) {
        self.order_book = OrderBook::new(self.initial_price);
        self.price_history = vec![self.initial_price];
        self.fundamental_history = vec![self.fundamental_value];
        self.trade_count = 0;
        self.current_step = 0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::agents::{ValueInvestor, MomentumTrader, MarketMaker};

    #[test]
    fn test_simulation_runs() {
        let mut engine = SimulationEngine::new(100.0, 100.0, 0.02);

        engine.add_agent(Box::new(ValueInvestor::new(
            "value_1".to_string(),
            100000.0,
            100,
            100.0,
        )));

        engine.add_agent(Box::new(MarketMaker::new(
            "mm_1".to_string(),
            200000.0,
            200,
        )));

        let result = engine.run(100);

        assert_eq!(result.steps_completed, 100);
        assert!(result.price_history.len() > 100);
        assert_eq!(result.agent_results.len(), 2);
    }

    #[test]
    fn test_price_changes() {
        let mut engine = SimulationEngine::new(100.0, 100.0, 0.05);

        engine.add_agent(Box::new(MomentumTrader::new(
            "momentum_1".to_string(),
            100000.0,
            50,
        )));

        let result = engine.run(50);

        // Price should change from initial
        let min_price = result.price_history.iter().cloned().fold(f64::INFINITY, f64::min);
        let max_price = result.price_history.iter().cloned().fold(f64::NEG_INFINITY, f64::max);

        assert!(max_price > min_price);
    }
}
